### DISCUSSION TITLE
DecisionTreeRegressor -ml_algo.(kaggle_doubt)

### DISCUSSION THREAD

**Discussion by AKA-RONY** [dsc-tiu/MLStudyJams2022#10] [Q&A] [answered]
https://www.kaggle.com/code/dansbecker/your-first-machine-learning-model?scriptVersionId=79125364&cellId=15
doubt1:-
in continuation of the above course, in **Building Your model** Section, we have first defined a model, so what's that (random_state=1) means? 

doubt2:-
also after fitting the model with (x,y) values, and then running the predict function-
print(melbourne_model.predict(X.head())) #how the predict function works,we just mentioned the X parameter i.e the features which will be used for prediction, but we didn't mentioned the y parameter i.e the target variable(salesprice). Is it like the predict function is too smart that it knows which target variable we are referring to and give the predictions  based on the features passed? or something else> pls reply

**Comment by diffrxction** [comment, marked as answer]
@salvik43 Hey Salvik! 

For doubt 1: random_state is a parameter that ensures that the data is split and sent into the machine learning model in a specific way. The way it helps us is that if two different ML models were being trained using the same algorithm and data, then setting the same random_state would ensure same portions of the data are being pushed into the model and would ensure very similar results in both cases. This could help us shift our focus to other parameters passed when training or initializing the model and we could adjust those to fine tune our results.

For doubt 2: predict function is called once the model has been completely trained with the training set. The goal of the predict function is to take in testing data and generate results. Now these generated results can be compared with original and true results within the data set and this allows us to see the difference in the results and use formulas like those of accuracy, recall, precision and f1 to correctly deduce how well the model performs.
Predict also helps make a prediction on new data and once the model accuracy is satisfactory, we can use it to gain predictions on the future if we have the relevant data.

**Reply by NeelaGitH** [reply]
To split data into training and testing data, we need to call the train_test_split function. So I can understand why we need a random_state parameter in the train_test_split function. But I guess DecisionTreeRegressor is just for defining the model and it can't split the data by itself, so why do we need the random_state parameter there? What's the specific purpose of the random_state argument there? 
I hope I made myself clear.

**Reply by AKA-RONY** [reply]
When you split your data into training and testing sets using train_test_split, the random_state parameter is used to initialize the random number generator used for shuffling the data before splitting it. If you don't set a random_state, the data will be split differently every time you run your code, which could lead to inconsistent results.

Similarly, when you train a DecisionTreeRegressor model, the algorithm randomly selects features to split on at each node. The random_state parameter in DecisionTreeRegressor is used to set the seed for the random number generator used for selecting the features. Setting a random_state for the model ensures that the same features are selected at each node when the model is trained multiple times on the same data.
hope this helps...

**Reply by NeelaGitH** [reply]
Hey, thanks for the explanation. But I have some doubts. Can you elaborate on what you mean by "used to set the seed for the random number generator"? I mean, elaborate the word seed a bit if you can.

And basically random_state parameter in train_test_split func makes sure the data is shuffled and split the same way every time,right?

**Reply by NeelaGitH** [reply]
> To split data into training and testing data, we need to call the train_test_split function. So I can understand why we need a random_state parameter in the train_test_split function. But I guess DecisionTreeRegressor is just for defining the model and it can't split the data by itself, so why do we need the random_state parameter there? What's the specific purpose of the random_state argument there? 
> I hope I made myself clear.

This is a follow-up question regarding this-
 so why is the random_state parameter in the Decision Tree Regressor function given the same value as the random_state parameter in train_test_split function? What happens if they're given different values?

**Reply by diffrxction** [reply]
> Hey, thanks for the explanation. But I have some doubts. Can you elaborate on what you mean by "used to set the seed for the random number generator"? I mean, elaborate the word seed a bit if you can.
> 
> And basically random_state parameter in train_test_split func makes sure the data is shuffled and split the same way every time,right?

Basically the seed is a fixed starting point for a whole sequence of random numbers. When you set a seed this means you are ensuring a very specific starting point for the RNG used by the model. In a more abstract way think of it like this, model is trying to find relation between input and output (considering data is labeled and problem solution is supervised learning), so the model has to start at a random relation between input and output, this relation is mathematical and hence expressed as a sequence of random numbers, over a course (epochs) of training, the model then adjusts those random numbers to form the most correct relation between input and output. 

RNG simply serves as a starting point for the model to get a sequence of random numbers to act as the base from where improving the model or 'machine learning' can take place. Hope this helps!

**Reply by diffrxction** [reply]
> > To split data into training and testing data, we need to call the train_test_split function. So I can understand why we need a random_state parameter in the train_test_split function. But I guess DecisionTreeRegressor is just for defining the model and it can't split the data by itself, so why do we need the random_state parameter there? What's the specific purpose of the random_state argument there?
> > I hope I made myself clear.
> 
> This is a follow-up question regarding this- so why is the random_state parameter in the Decision Tree Regressor function given the same value as the random_state parameter in train_test_split function? What happens if they're given different values?

you can have a completely different number if you want to, both are separate functions with separate use cases (one for data spliting and one for algorithm's sequence of random numbers) and their RNGs behavior is not mutual.
